src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1248,6 +1220,123 @@ private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
             }
         }
 
+        private bool IsRefStruct(out bool isRefFollowedByStructOrPartialStruct, out bool isTypeDeclaration)
+        {
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.RefKeyword);
+
+            var nextToken = PeekToken(1);
+
+            isRefFollowedByStructOrPartialStruct = nextToken.Kind == SyntaxKind.StructKeyword ||
+                (nextToken.ContextualKind == SyntaxKind.PartialKeyword && this.PeekToken(2).Kind == SyntaxKind.StructKeyword);
+
+            if (isRefFollowedByStructOrPartialStruct)
+            {
+                isTypeDeclaration = true;
+                return true;
+            }
+
+            isTypeDeclaration = IsTypeDeclaration(out nextToken);
+            return isTypeDeclaration && nextToken.Kind == SyntaxKind.StructKeyword;
+        }
+
+        private static bool CanFollowTypeName(SyntaxKind kind)
+        {
+            switch (kind)
+            {
+                case SyntaxKind.DotToken:                   // partial.     qualified
+                case SyntaxKind.ColonColonToken:            // partial::    alias
+                case SyntaxKind.LessThanToken:              // partial<     generic type
+                case SyntaxKind.OpenBracketToken:           // partial[     array
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static bool CanFollowMemberName(SyntaxKind kind)
+        {
+            switch (kind)
+            {
+                case SyntaxKind.LessThanToken:              // partial<     generic method
+                case SyntaxKind.OpenBraceToken:             // partial {    property
+                case SyntaxKind.EqualsGreaterThanToken:     // partial =>   property
+                case SyntaxKind.OpenParenToken:             // partial(     method
+                    return true;
+            }
+
+            return false;
+        }
+
+        private enum PartialModFlags
+        {
+            PartialType,
+            PartialMember,
+            TreatAsModifier,
+            NotModifier
+        }
+
+        private PartialModFlags ShouldPartialBeTreatedAsModifier()
+        {
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.PartialKeyword);
+
+            int peekIndex = 1;
+            SyntaxToken currentToken;
+
+            while (GetModifier(currentToken = PeekToken(peekIndex)) != DeclarationModifiers.None)
```

Comment:
this was so confusing to read until i *finally* realizing you're assigning inside an argument.  Can we just actually break this out into individual variables and flow control?